tensorflow/core/kernels/sparse_to_dense_op_gpu.cu.cc

```diff
@@ -0,0 +1,186 @@
+    bool valid = true;
+
+    if (thread_idx == 0) {
+      for (int di = 0; di < ndims; di++) {
+        Index curr_idx = indices[thread_idx * ndims + di];
+        if (curr_idx < 0 || curr_idx >= dims[di]) valid = false;
+      }
+      different = true;
+    } else {
+      for (int di = 0; di < ndims; di++) {
+        Index curr_idx = indices[thread_idx * ndims + di];
+        if (curr_idx < 0 || curr_idx >= dims[di]) valid = false;
+        Index prev_idx = indices[(thread_idx - 1) * ndims + di];
+        Index diff = curr_idx - prev_idx;
+        if (diff > 0) different = true;
+        if (!different && diff < 0) increasing = false;
+      }
+    }
+
+    if (!valid) {
+      atomicMin(detail::ToCudaSupportedPtr(&status[0]), thread_idx);
+    }
+    if (!increasing) {
+      atomicMin(detail::ToCudaSupportedPtr(&status[1]), thread_idx);
+    }
+    if (!different) {
+      atomicMin(detail::ToCudaSupportedPtr(&status[2]), thread_idx);
+    }
+  }
+}
+
+// IndicesValidStatus contains three status for the out-of-bound check, the
+// sorted check, and the repeat check. If the value equals to INT_MAX, the
+// check passes. Otherwise, it represents the first detected position of the
+// invalid index for the check.
+struct IndicesValidStatus {
+  int valid;
+  int increasing;
+  int different;
+};
+
+}  // namespace
+
+namespace functor {
+
+template <typename T, typename Index>
+void LaunchSparseToDense<T, Index>::operator()(
+    OpKernelContext* c, bool validate_indices, const Index* indices,
+    const T* values, const int num_elems, const int num_values,
+    const Index* shape, const int num_dims, const T default_value,
+    Index dense_size, T* dense) {
+  auto* stream = c->op_device_context()->stream();
+  const Eigen::GpuDevice& d = c->eigen_gpu_device();
+
+  if (validate_indices && num_elems != 0) {
+    LOG(WARNING) << "SparseToDense will be performed on GPUs. For performance "
+        "reasons, it is suggested to pass False to validate_indices.";
+
+    IndicesValidStatus valid_status;
+    int valid_status_size = sizeof(valid_status) / sizeof(int);
+    int valid_status_bytes = sizeof(valid_status);
+
+    Tensor valid_status_tensor;
+    OP_REQUIRES_OK(c, c->allocate_temp(
+        DT_INT32, TensorShape({valid_status_size}),
+        &valid_status_tensor));
+
+    auto status_ptr = valid_status_tensor.template flat<int>().data();
+    se::DeviceMemoryBase valid_status_ptr(status_ptr, valid_status_bytes);
+
+    GpuLaunchConfig config = GetGpuLaunchConfig(num_elems, d);
+    stream->ThenMemset32(&valid_status_ptr, INT_MAX, valid_status_bytes);
+    OP_REQUIRES_OK(c, GpuLaunchKernel(
+        CheckIndicesValid<Index>, config.block_count, config.thread_per_block,
+        0, d.stream(), indices, num_elems, shape, num_dims, status_ptr));
+    stream->ThenMemcpy(reinterpret_cast<int*>(&valid_status), valid_status_ptr,
+                       valid_status_bytes);
+    OP_REQUIRES_OK(c, stream->BlockHostUntilDone());
```

Comment:
As stated in the header, this should use a callback on the `EventMgr` and the `OpKernel` should be an `AsyncOpKernel`.